opm/grid/common/WellConnections.cpp

```diff
@@ -236,18 +271,41 @@ postProcessPartitioningForWells(std::vector<int>& parts,
                     assert(exportCandidate != exportList.end() && std::get<0>(*exportCandidate) == *movedCell);
                     std::get<1>(*exportCandidate) = new_owner;
                 }
-                owner = new_owner;
             }
-
-            well_indices_on_proc[owner].push_back(well_index);
-            ++well_index;
         }
         auto sorter = [](std::pair<const int, std::vector<int>> &pair) {
                           auto &vec = pair.second;
                           std::sort(vec.begin(), vec.end());
                       };
         std::for_each(addCells.begin(), addCells.end(), sorter);
         std::for_each(removeCells.begin(), removeCells.end(), sorter);
+
+        // prevent memory allocation
+        for (auto& well_indices : well_indices_on_proc) {
+            well_indices.reserve(wells.size());
+        }
+
+        // Check that all connections of a well have ended up on one process.
+        // If that is not the case for well then move them manually to the
+        // process that already has the most connections on it.
+        std::size_t well_index = 0;
+
+        for (const auto& well : wells) {
+            const auto& connections = well_connections[well_index];
+            if (connections.empty()) {
+                // Well does not perforate cells, nothing to worry about.
+                ++well_index;
+                continue;
+            } else {
+                int new_owner = parts[*connections.begin()];
+                well_indices_on_proc[new_owner].push_back(well_index);
+                const auto& old_owners_well = old_owners[well_index];
+                if (old_owners_well.size() > 1 || old_owners_well.find(new_owner) == old_owners_well.end()) {
+                    std::cout << "Manually moved well " << well.name() << " to partition " << new_owner << std::endl;
```

Comment:
This will not show up in the log, only on the terminal as it flies by. I suggest OpmLog::info() or debug().

Reply:
done